### PR TITLE
Automatically Generate Repository Git Page

### PR DESCRIPTION
**Pull request by bshifaw**
Added git action to generate wdl documentation and place them in gh-page branch
- wdldoc : creates markdown files for workflows        
- mkdocs: builds HTML site using markdown files.  

Added scripts to generate dot png files for pipeline wdls. 
- ./scripts/git_page/create_wdl_png.py 
- ./scripts/git_page/wdlviz.py (copied from miniwdl repo)

Added script to append png links to wdl markdown documentation   
- python3 ./scripts/git_page/add_dot_link_to_md.py

### CODE REVIEW

File: .github/workflows/git_page.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      - name: Setup Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.10'
+
+      - name: Install Requirements
+        run: |
+          python3 -m pip install --upgrade pip
+          python3 -m pip install wdldoc~=1.7.0
+          python3 -m pip install mkdocs
```
Comment (SHuang-Broad) [inline .github/workflows/git_page.yml:31]: pin down version

File: mkdocs.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+site_name: "Broad Institute Long Read Pipelines"
+repo_url: "https://github.com/broadinstitute/long-read-pipelines"
+
+site_description: "This repository contains pipelines for processing of long read data from PacBio and/or Oxford Nanopore platforms."
+site_author: "Broad Institute"
+copyright: "Copyright (c) 2019, Broad Institute"
```
Comment (SHuang-Broad) [inline mkdocs.yml:6]: 2019?

File: scripts/git_page/wdlviz.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    parser.add_argument("--inputs", action="store_true", help="include input declarations")
+    parser.add_argument("--outputs", action="store_true", help="include output declarations")
+    parser.add_argument("--output_path", help="Path to output directory")
+    parser.add_argument(
+        "--no-quant-check",
+        dest="check_quant",
+        action="store_false",
+        help="relax static typechecking of optional types, and permit coercion of T to Array[T] (discouraged; for backwards compatibility with older WDL)",
+    )
+    parser.add_argument(
```
Comment (SHuang-Broad) [inline scripts/git_page/wdlviz.py:36]: support option of choice between `svg` and `png`

File: scripts/git_page/utility.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import glob
+import logging
+from pathlib import Path, PurePosixPath, PurePath
+
+Logger = logging.getLogger(__name__)
+
+
+def get_absolute_path(path) -> Path:
```
Comment (SHuang-Broad) [inline scripts/git_page/utility.py:8]: let's typehint `path: str`

File: scripts/git_page/utility.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Get all files with a given extension in a directory
+    @param directory: directory to search
+    @param ext: extension (e.g. 'md', 'dot.png')
+    @return: list of files with extension
+    """
+    check_dir_exists(directory)
+
+    Logger.debug(f'Getting all files with extension {ext} in {directory}...')
+    return glob.glob(f'{directory}/**/*.{ext}', recursive=True)
```
Comment (SHuang-Broad) [inline scripts/git_page/utility.py:64]: this has one small bug, it will miss all the, say `.wdl` files, directly under `directory`.
Should be trivial to fix.

**Review by SHuang-Broad — CHANGES_REQUESTED** [review]
One tiny bug, good otherwise.
Also please don't forget about updating the Readme with a link to the documentation page.

File: .github/workflows/git_page.yml (hunk 1 above)
Comment (bshifaw) [reply .github/workflows/git_page.yml:31]: done

File: mkdocs.yml (hunk 2 above)
Comment (bshifaw) [reply mkdocs.yml:6]: done

File: scripts/git_page/utility.py (hunk 4 above)
Comment (bshifaw) [reply scripts/git_page/utility.py:8]: done

File: scripts/git_page/wdlviz.py (hunk 3 above)
Comment (bshifaw) [reply scripts/git_page/wdlviz.py:36]: done

File: scripts/git_page/utility.py (hunk 5 above)
Comment (SHuang-Broad) [reply scripts/git_page/utility.py:64]: Oh that's nice.

**Comment by bshifaw** [comment]
> One tiny bug, good otherwise. Also please don't forget about updating the Readme with a link to the documentation page.

I'd like to wait before adding in the pipeline site link in the readme. Just to be sure it works before broadcasting

File: scripts/git_page/md_post_process.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Performs post processing on markdown files. - Adds title header to markdown files if it does not "
+    )
+    parser.add_argument(
+        "--md_dir", metavar="MARKDOWN_DIRECTORY", required=True,
+        help="One of or more directories containing markdown "
+             "documentation files for workflow"
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:19]: how to specify the multiple directories? A ` ` separated list?

File: scripts/git_page/md_post_process.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    )
+    parser.add_argument("--debug", action="store_true", help="verbose logging")
+
+    args = parser.parse_args()
+
+    util.set_logging_level(args)
+
+    md_dir_name = args.md_dir
+
+    md_dir = util.get_absolute_path(md_dir_name)
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:29]: It looks like `md_dir_name` is a single dir, not the multiple folders as indicated in the doc?

File: scripts/git_page/md_post_process.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    the word 'workflows'
+    :param md_path: Path to markdown file
+    :return: True if the markdown file path contains 'workflows'
+    """
+    return "/workflows/" in md_path.lower()
+
+
+def workflow_md_has_multiple_subheaders(md_path: str) -> bool:
+    """
+    Checks if the markdown file has multiple subheaders
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:93]: What does multiple subheaders indicate?

File: scripts/git_page/md_post_process.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+def workflow_md_has_task_subheader(md_path: str) -> bool:
+    """
+    Checks if the markdown file has a task subheader
+    :param md_path: Path to markdown file
+    :return: True if the markdown file has a task subheader
+    """
+    try:
+        with open(md_path, 'r') as md_file:
+            md_lines = md_file.readlines()
+            if len(md_lines) == 0:
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:177]: lines 177-182 can be out of the `with` block.

File: scripts/git_page/md_post_process.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+        subheader_count = 0
+        for line in md_lines:
+            if line.startswith('## '):
+                subheader_count += 1
+
+        return subheader_count > 1
+
+
+def add_task_subheader(md_path: Union[str, Path]) -> None:
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:110]: Is this function used? If not, let's take it away.

File: scripts/git_page/md_post_process.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    md_paths = util.get_all_files_with_extension(directory=md_dir, ext='md')
+    Logger.debug(f'Found {len(md_paths)} markdown files')
+
+    for md in md_paths:
+        Logger.debug(f'Processing {md}...')
+        if not check_if_md_file_has_title_header(md_path=md):
+            Logger.debug(f'Adding title header to {md}')
+            add_title_header(md_path=md)
+
+        if md_path_in_workflows(md_path=md) and workflow_md_has_multiple_subheaders(md_path=md):
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:40]: Why is this necessary?

File: scripts/git_page/md_post_process.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                return
+            md_file.seek(0)
+            md_file.truncate()
+            md_file.writelines(lines[:second_subhead_index])
+
+    except (FileNotFoundError, IsADirectoryError):
+        Logger.debug(f"Unable to open file at {md_path}")
+
+
+def workflow_md_has_task_subheader(md_path: str) -> bool:
```
Comment (SHuang-Broad) [inline scripts/git_page/md_post_process.py:168]: function appear unused?

File: scripts/git_page/md_post_process.py (hunk 7 above)
Comment (bshifaw) [reply scripts/git_page/md_post_process.py:29]: done

File: scripts/git_page/md_post_process.py [hunk 13]
```diff
@@ -117,24 +117,25 @@ def remove_subheaders_after_first(md_path: Union[str, Path]) -> None:
     """
     try:
         with Path(md_path).open('r+') as md_file:
-            lines = [s.strip() for s in md_file.readlines()]
```
Comment (bshifaw) [inline scripts/git_page/md_post_process.py]: the contents of `lines` will be used to rewrite the file below so the `\n` should be stripped away

File: scripts/git_page/md_post_process.py (hunk 9 above)
Comment (bshifaw) [reply scripts/git_page/md_post_process.py]: done

File: scripts/git_page/md_post_process.py (hunk 12 above)
Comment (bshifaw) [reply scripts/git_page/md_post_process.py]: removed
